### ISSUE TITLE
Branching and Release Strategy going forward

### ISSUE THREAD

**Issue by trainerbill** [meanjs/mean#757]
We have been talking about a new branching/release strategy in the chat for a couple days now.  Bringing this to an issue for additional community feedback.

@rhutchison @mleanos @codydaig @lirantal @ilanbiala 

Here was my thoughts:

In my mind master is always working to the 0.x.0 release and that development starts the second we release 0.4.0. I don't want to wait on my changes. So when 0.4.0 is released we make a branch 0.4.x and tag 0.4.0. Now bug fixes go to 0.4.x branch and tag 0.4.1. The bug fix is added to the 0.5.0 milestone so that someone goes into master to see if the bug still exists in master and if so fix it. Now we have a list of bug fixes from the release branches that we create one PR to fix them all in master. Just do that at certain intervals to close out "bug fixes" from the milestone. When 0.5.0 milestone has been reached we make branch 0.5.x and tag 0.5.0 and proceed the same. Bug fixes to 0.5.x with an incremented tag and new feature/changes to master working toward 0.6.0 milestone.

All new development once 0.4.0 is released goes to master on the 0.5.0 milestone. This would include like the unauthorized PR and Pagination PR that I submitted. Only bug fixes would go to a release branch, 0.4.x or 0.3.x, and when we merge a bug fix PR to a version branch we should be tagging the branch with an incremented patch version and maintaining a bug fix section in the next milestone to make sure that the fix that was submitted to the version branch is not in master.

It follows the angularjs repo as they do pretty much the same thing.

**Comment by codydaig** [comment]
Perfect.

**Comment by trainerbill** [comment]
And before @lirantal asks what issues this addresses:

1.  Releases are super simple.  Just git tag.
2.  PR's will not have to be closed and reopened on another branch when we release a new version since everything is going to master.

**Comment by ilanbiala** [comment]
@trainerbill I like the idea, but the only thing I think will be different from the angularjs repo is that I don't think we should deal with the complexity of 2 version branches at once. AngularJS has to do this because it needs to support users for a while until upgrading is actually logical when the features and specs are commonplace.

**Comment by lirantal** [comment]
I'll just add that I'm not really in favor of those bug branches that @trainerbill is describing because this begins again the branch management over-head.

Basically I see it like this:
1. We are merging PRs to `master` branch
2. When we achieve some milestone and wish to release 0.5.0 we tag the `master` branch with `v0.5.0` tag 
3. Goto (1)

@trainerbill regarding
```
So when 0.4.0 is released we make a branch 0.4.x and tag 0.4.0. Now bug fixes go to 0.4.x branch and tag 0.4.1.
````
Why do we need to create a 0.4.x branch after 0.4.0 is released as a tag? After a release was made (basically that's a tag) *any* new development (PR) is merged to master for the next release.
If you are referring for some way to back-port bug fixes made on master to previous releases then I think that this will require extreme branch management and it doesn't work the investment required. If someone needs a bugfix for something they either: 1. upgrade to another version where the issue is fixed, or upgrade to `master` branch; 2. they back-port the fix themselves.

**Comment by codydaig** [comment]
@lirantal I agree. The bug branches only work if we're going to support previous versions, and it doesn't sound like we will, so it's simplest to keep everything in master and just tag as we go! :-D

**Comment by trainerbill** [comment]
@lirantal @codydaig If you are not planning on supporting previous releases then that is fine but then we cannot increment patch versions, ie 0.4.1, as tagging the master branch at 0.4.1 may include commits that are incompatible with 0.4.0.  I don't think that supporting one previous minor version 0.x.0 would be a huge undertaking.  Thats why I like the idea of a 0.x.x branch that gets fixes outside of master as master is rolling towards 0.x.0 release and contains incompatible code.  Or you say that patch versions are incompatible with prior releases which I don't think people will understand.  I would assume if i switch from 0.4.0 to 0.4.1 that ALL my custom modules would work as they did in 0.4.0.  

Having a list of bug fix PRs that went into release branches on the milestone would be easy.  PR is merged to 0.4.x and you just add the PR to the 0.5.0 milestone so when the time comes to release 0.5.0 we stop feature development and go through the list of bug fixes on 0.4.x and make sure they are not present in 0.5.0.  I would imagine this would be done very quickly.

**Comment by rkoberg** [comment]
Perhaps look into using gitflow: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

"This workflow doesn’t add any new concepts or commands beyond what’s required for the Feature Branch Workflow. Instead, it assigns very specific roles to different branches and defines how and when they should interact. In addition to feature branches, it uses individual branches for preparing, maintaining, and recording releases. Of course, you also get to leverage all the benefits of the Feature Branch Workflow: pull requests, isolated experiments, and more efficient collaboration."

**Comment by lirantal** [comment]
Again, as far as I see it, it's binary:
1. We either simplify this process altogether: one `master` branch, and tagging releases
2. We keep the current version-work branches

IMHO If we try to (1) with anymore branch handling this will really make it hard, especially when me and Ilan are the only users whom are able to assign/merge PRs (I pinged Roie about this issue to see if we can have more freedom in adding more contributors but for now no response yet).

@trainerbill about not breaking an 0.4.0 with a future 0.4.2 - well that's part of the version plan and we should all be aware when we consider PRs and perform code review that if something is being added and contains risks to break something in that 0.x.0 version then it should be made very clear and at that point we decide to either:
1. postpone that PR to a later time until we are finish with the current 0.x.0 version
2. we want that PR in so in order to accept it we jump an 0.x.0 (i.e: from 0.4.0 to 0.5.0)
Is that better explained? do you agree?

**Comment by jloveland** [comment]
I use git flow in the office and it enables us to manage releases and hotfixes very easily. Everyone develops features off of the develop branch. Master is always deployable to customers and versions are tagged here. We can have multiple customers on different versions. It takes a little getting used to because everyone works off of develop, but the git-flow command line tool automates merging across the different branches, hotfixes, releases, feature branches, etc..

**Comment by lirantal** [comment]
We're using the git successful model here http://nvie.com/posts/a-successful-git-branching-model/ along with Stash, and Pull-Requests, code reviews, etc.
But what I'm saying is that the nature of office work where you version product and the amount of resources, and schedules you have is simply *not the same* as you have with managing open source projects where things are much more dynamic.

I think we're overall in alignment with regards to merging now to master and we've already started doing that so this is good. Let's see how that goes and if we'll really need any other branches in the future.

So I guess this sums it up...?

**Comment by ilanbiala** [comment]
@trainerbill I'm not really in favor of supporting previous releases, and here's my philosophy:

I'm working on a project that is based on the framework, I'm not going to spend as much time upgrading MEAN every few months, maybe more frequently, as I do coding because then I'm not adding as many features to my project. Rather, I'm moving around files and getting stuff like chat modules when in reality I don't need that in most cases. When I start a new project, I'll just use the newest version of MEAN and enjoy the changes.

Maybe it's just me, but most of the time MEAN is improving on stuff that is scaffolding and not actually going to help an in-progress project improve compared to the improvement noticed when starting off a new project with an up-to-date version of MEAN.

**Comment by mleanos** [comment]
I prefer the Feature Branch Flow, and I use `git flow` for my projects. However, if the consensus here is that @trainerbill suggestion is going to be most beneficial then let's roll with it.

My main concern, during the duration of this debate, has been the ease in which other developers can come into the project and start contributing; without having to know what our exact branching/tagging strategy is.

From what @trainerbill @rhutchison @codydaig are saying, it will be easy. I trust their judgement, and I'm willing to follow their lead on this; even if it will be a little uncomfortable for me at first.

**Comment by mleanos** [comment]
@ilanbiala I feel you on that. And it's an interesting perspective. I have an ongoing project that is based off this framework. I upgraded to the 0.4.0 branch a few months ago, and have been keeping it in sync as much as possible. Usually, when I'm done with a branch/feature/bug-fix on my own project. I'm trying to be as active as I can in this community, but I think you have a very valid point. My project isn't really going to benefit that much by me constantly keeping it in sync with MEAN. Or for anyone one that doesn't have the time to keep up with the changes, and doesn't really need the newest version that often.

I'm not sure how much the proposed strategy would affect these concerns though. It seems that someone getting started with the MEAN stack would just get the latest version & run with it. Even if, technically, the master branch isn't the most "stable" version.

**Comment by trainerbill** [comment]
@ilanbiala You are probably right.  And the dev can always just go in and fix the bug on their own.  I am ok with pr'ing to master for now without release branches.  If we need to support previous releases we can revisit this in the future.

@lirantal 
1. postpone that PR to a later time until we are finish with the current 0.x.0 version
2. we want that PR in so in order to accept it we jump an 0.x.0 (i.e: from 0.4.0 to 0.5.0)
Number one is what I am trying to solve for!  Number two would make us switch versions way to often.  I say we just commit to master for now and put a warning in the documentation.

@codydaig I say that we just make it very clear in the documentation that master is to be used at your own risk.  If you want a stable version you need to checkout a release tag.

**Comment by codydaig** [comment]
@trainerbill if you can, submit pull request or open an issue and tag me. i wont have time until next week. 

@ilanbiala well said

**Comment by lirantal** [comment]
Ok, so this is concluded.
@ilanbiala's description is another argument for working with a master branch, and no other special releases/bugfix branches.

**Comment by codydaig** [comment]
@lirantal I'll update the contributing document I submitted a pull request for to reflect that.

**Comment by lirantal** [comment]
Great, thanks Cody
